### PR TITLE
Add support for vision chat completion models

### PR DESCRIPTION
**Pull request by pufferffish**
**Describe the change**
This PR let's you pass in `image_url` so we can pass in images to vision models like `gpt-4-vision-preview`

**Describe your solution**
The definition of `ChatCompletionMessage` is changed such that we can pass in a struct containing text and images for `Content` instead of a string.

**Tests**
This PR has been tested in `example/chatvision`

Issue: #539

### CODE REVIEW

File: chat.go [hunk 1]
```diff
@@ -51,7 +56,31 @@ type PromptAnnotation struct {
...
+
+type ChatMessageContent struct {
+	Type     string               `json:"type"`
+	Text     string               `json:"text,omitempty"`
+	ImageURL *ChatMessageImageURL `json:"image_url,omitempty"`
+}
+
 type ChatCompletionMessage struct {
+	Role    string               `json:"role"`
+	Content []ChatMessageContent `json:"content"`
```
Comment (rkintzi) [inline chat.go:71]: This decision is not within my jurisdiction, but these modifications breaks compatibility. I have already suggested a less intrusive alternative [here](https://github.com/sashabaranov/go-openai/issues/539#issuecomment-1799826250). I am awaiting the maintainer's involvement to facilitate a discussion on the matter.

Comment (zquestz) [reply chat.go:71]: The OpenAI API still accepts string values for Content. Ideally this wrapper conforms to their API contract.

File: chat.go [hunk 2]
```diff
@@ -51,7 +56,31 @@ type PromptAnnotation struct {
 	ContentFilterResults ContentFilterResults `json:"content_filter_results,omitempty"`
 }
 
+type ChatMessageImageURL struct {
+	URL string `json:"url"`
```
Comment (AlexandrosKyriakakis) [inline chat.go:60]: You can also add detail attribute. https://platform.openai.com/docs/guides/vision/calculating-costs

File: examples/chatvision/main.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+					{
+						Type: openai.ChatMessageContentTypeImage,
+						ImageURL: &openai.ChatMessageImageURL{
+							URL: "https://upload.wikimedia.org/wikipedia/commons/thumb/d/dd/Gfp-wisconsin-madison-the-nature-boardwalk.jpg/2560px-Gfp-wisconsin-madison-the-nature-boardwalk.jpg",
+						},
+					},
+				},
+			},
+		},
+	}
```
Comment (AlexandrosKyriakakis) [inline examples/chatvision/main.go:33]: Can also include an example with local image like this one.

```go
	imagePath := "path/to/image.png"

	// Read the image file
	imgData, err := os.ReadFile(imagePath)
	if err != nil {
		fmt.Println("Error reading image file:", err)
		os.Exit(1)
	}

	// Encode to base64
	base64Str := base64.StdEncoding.EncodeToString(imgData)
	req := openai.ChatCompletionRequest{
		Model:     openai.GPT4VisionPreview,
		MaxTokens: 4096,
		Messages: []openai.ChatCompletionMessage{
			{
				Role: openai.ChatMessageRoleUser,
				Content: []openai.ChatMessageContent{
					{
						Type: openai.ChatMessageContentTypeText,
						Text: "What's in this image",
					},
					{
						Type: openai.ChatMessageContentTypeImage,
						ImageURL: &openai.ChatMessageImageURL{
							Detail: openai.ImageURLDetailHigh,
							URL:    "data:image/png;base64," + base64Str,
						},
					},
				},
			},
		},
	}
```

File: chat.go (hunk 2 above)
Comment (AlexandrosKyriakakis) [reply chat.go:60]: I made it like this
```go
type ImageURLDetail string

const (
	ImageURLDetailLow  ImageURLDetail = "low"
	ImageURLDetailHigh ImageURLDetail = "high"
)

type ChatMessageImageURL struct {
	Detail ImageURLDetail `json:"detail"`
	URL    string         `json:"url"`
}
```

File: chat.go (hunk 1 above)
Comment (rkintzi) [reply chat.go:71]: @sashabaranov 
Do you have any idea how to deal with backward compatibility? I summed-up the possible solutions (that I see) here.

**Comment by arif599** [comment]
Hey! is it possible to stream with gpt-4-vision-preview?

**Comment by arif599** [comment]
> Hey! is it possible to stream with gpt-4-vision-preview?

I think I figured it out. However, the stream sometimes receives EOF during middle of the sentence.
'''
package main

import (
	"context"
	"errors"
	"fmt"
	"io"
	"os"

	openai "github.com/sashabaranov/go-openai"
)

func main() {
	client := openai.NewClient(os.Getenv("OPENAI_API_KEY"))
	ctx := context.Background()

	req := openai.ChatCompletionRequest{
		Model: openai.GPT4VisionPreview,
		Messages: []openai.ChatCompletionMessage{
			{
				Role: openai.ChatMessageRoleUser,
				Content: []openai.ChatMessageContent{
					{
						Type: openai.ChatMessageContentTypeText,
						Text: "What's in this image",
					},
					{
						Type: openai.ChatMessageContentTypeImage,
						ImageURL: &openai.ChatMessageImageURL{
							URL: "https://upload.wikimedia.org/wikipedia/commons/thumb/d/dd/Gfp-wisconsin-madison-the-nature-boardwalk.jpg/2560px-Gfp-wisconsin-madison-the-nature-boardwalk.jpg",
						},
					},
				},
			},
		},
		Stream: true,
	}

	stream, err := client.CreateChatCompletionStream(ctx, req)
	if err != nil {
		fmt.Printf("ChatCompletionStream error: %v\n", err)
		return
	}
	defer stream.Close()

	fmt.Printf("Stream response: ")
	for {
		response, err := stream.Recv()
		if errors.Is(err, io.EOF) {
			fmt.Println("\nStream finished")
			return
		}

		if err != nil {
			fmt.Printf("\nStream error: %v\n", err)
			return
		}

		fmt.Printf(response.Choices[0].Delta.Content)
	}
}
'''

**Comment by rkintzi** [comment]
A few days ago, I encountered the same issue. Typically, I receive a sentence and a half, regardless of whether I send a URL with the https:// prefix or with an embedded image. I haven't tried without streaming enabled but will do so in a few hours. Have you noticed any differences in results when using streaming versus not using it?

@arif599

**Comment by AlexandrosKyriakakis** [comment]
@arif599 I believe you need to set the MaxTokens to be higher value or highest(= 4096).
I guess it similar to this issue -> https://community.openai.com/t/gpt-4-vision-preview-finish-details/475911

**Comment by arif599** [comment]
> @arif599 I believe you need to set the MaxTokens to be higher value or highest(= 4096). I guess it similar to this issue -> https://community.openai.com/t/gpt-4-vision-preview-finish-details/475911

Added `MaxTokens: 4096` in ChatCompletionRequest and is working fine, thanks :)

**Comment by arif599** [comment]
> A few days ago, I encountered the same issue. Typically, I receive a sentence and a half, regardless of whether I send a URL with the https:// prefix or with an embedded image. I haven't tried without streaming enabled but will do so in a few hours. Have you noticed any differences in results when using streaming versus not using it?
> 
> @arif599

Haven't noticed any difference with the quality of results but feel free to test it out yourself. 
For streaming, just add MaxTokens: 4096 to ChatCompletionRequest in the example I provided.
So now it would be 


`package main

import (
	"context"
	"errors"
	"fmt"
	"io"
	"os"

	openai "github.com/sashabaranov/go-openai"
)

func main() {
	client := openai.NewClient(os.Getenv("OPENAI_API_KEY"))
	ctx := context.Background()

	req := openai.ChatCompletionRequest{
		Model: openai.GPT4VisionPreview,
		Messages: []openai.ChatCompletionMessage{
			{
				Role: openai.ChatMessageRoleUser,
				Content: []openai.ChatMessageContent{
					{
						Type: openai.ChatMessageContentTypeText,
						Text: "What's in this image",
					},
					{
						Type: openai.ChatMessageContentTypeImage,
						ImageURL: &openai.ChatMessageImageURL{
							URL: "https://upload.wikimedia.org/wikipedia/commons/thumb/d/dd/Gfp-wisconsin-madison-the-nature-boardwalk.jpg/2560px-Gfp-wisconsin-madison-the-nature-boardwalk.jpg",
						},
					},
				},
			},
		},
		Stream:    true,
		MaxTokens: 4096,
	}

	stream, err := client.CreateChatCompletionStream(ctx, req)
	if err != nil {
		fmt.Printf("ChatCompletionStream error: %v\n", err)
		return
	}
	defer stream.Close()

	fmt.Printf("Stream response: ")
	for {
		response, err := stream.Recv()
		if errors.Is(err, io.EOF) {
			fmt.Println("\nStream finished")
			return
		}

		if err != nil {
			fmt.Printf("\nStream error: %v\n", err)
			return
		}

		fmt.Printf(response.Choices[0].Delta.Content)
	}
}
`

**Comment by arif599** [comment]
@rkintzi can we send base64 encoded imgs?
Like https://platform.openai.com/docs/guides/vision/uploading-base-64-encoded-images

I wrote this encodeImage function in go

```
func encodeImage(imagePath string) string {
	imageData, err := os.ReadFile(imagePath)
	if err != nil {
		fmt.Println("Error:", err)
		return ""
	}
	base64String := base64.StdEncoding.EncodeToString(imageData)
	return base64String
}
```
but when I pass the encoded img into ImgURL

```
	req := gpt.ChatCompletionRequest{
		Model: gpt.GPT4VisionPreview,
		Messages: []gpt.ChatCompletionMessage{
			{
				Role: gpt.ChatMessageRoleUser,
				Content: []gpt.ChatMessageContent{
					{
						Type: gpt.ChatMessageContentTypeText,
						Text: prompt,
					},
					{
						Type: gpt.ChatMessageContentTypeImage,
						ImageURL: &gpt.ChatMessageImageURL{
							URL: encodeImage("example.jpg"),
						},
					},
				},
			},
		},
		Stream:    true,
		MaxTokens: 4096,
	}
```

I get error, status code: 400, message: Invalid image.

**Comment by AlexandrosKyriakakis** [comment]
@sashabaranov @pufferffish closing this?

**Pull request closed without merging**